py3status/modules/google_calendar.py

```diff
@@ -0,0 +1,424 @@
+
+    Download the client_secret.json file which contains your client ID and
+    client secret. In your i3status config file, set configuration parameter
+    client_secret to the path to your client_secret.json file.
+
+    The first time you run the module, a browser window will open asking you
+    to authorize access to your calendar. After authorization is complete,
+    an access/refresh token will be saved to the path configured in
+    auth_token, and i3status will be restarted. This restart will
+    occur only once after the first time you successfully authorize.
+
+Example:
+```
+google_calendar {
+    client_secret = '/path/to/client_secret/'
+    auth_token = '/path/to/auth_token/'
+}
+```
+@author Igor Grebenkov
+@license BSD
+"""
+
+import httplib2
+import os
+import datetime
+
+from apiclient import discovery
+from oauth2client import client
+from oauth2client import tools
+from oauth2client.file import Storage
+from httplib2 import ServerNotFoundError
+
+
+SCOPES = 'https://www.googleapis.com/auth/calendar.readonly'
+APPLICATION_NAME = 'py3status google_calendar module'
+
+
+class Py3status:
+    auth_token = '~/.config/py3status/google_calendar.auth_token'
+    button_open = 3
+    button_toggle = 1
+    cache_timeout = 60
+    client_secret = '~/.config/py3status/google_calendar.client_secret'
+    format = '{events}|No Events'
+    format_date = '%a %d-%m'
+    format_error = ''
+    format_event = '\?color=threshold {summary} {time_to}'
+    format_event_full = '\?color=threshold {summary} ({start_time} - {end_time}, {start_date})'
+    format_separator = '\|'
+    format_time = '%I:%M %p'
+    format_time_left = '({days}d {hours}h {minutes}m) left'
+    format_time_to = '({days}d {hours}h {minutes}m)'
+    num_events = 3
+    thresholds = []
+    time_to_max = 180
+    warn_threshold = 0
+    warn_timeout = 300
+
+    def post_config_hook(self):
+        self.button = None
+        self.button_states = [False] * self.num_events
+        self.events = None
+        self.no_update = False
+
+        self.credentials = self._get_credentials()
+        self.is_authorized = self._authorize_credentials()
+
+    def _get_credentials(self):
+        """
+        Gets valid user credentials from storage.
+
+        If nothing has been stored, or if the stored credentials are invalid,
+        the OAuth2 flow is completed to obtain the new credentials.
+
+        Returns: Credentials, the obtained credential.
+        """
+        client_secret_path = os.path.dirname(os.path.expanduser(self.client_secret))
+        auth_token_path = os.path.dirname(os.path.expanduser(self.auth_token))
+
+        if not os.path.exists(auth_token_path):
+            os.makedirs(auth_token_path)
+
+        if not os.path.exists(client_secret_path):
+            os.makedirs(client_secret_path)
+
+        self.auth_token = os.path.expanduser(self.auth_token)
+        self.client_secret = os.path.expanduser(self.client_secret)
```

Comment:
we should do this in `post_config_hook()` then it will only be done once as we do this a few lines before here.

It also will make the code more readable imho

Reply:
done.

Reply:
I think he want you to move them up so we can `os.path.expanduser(word)` once in the beginning and we don't have to parse anything more than `word` inside `something()`.

Reply:
Oh, I see, 'cause I do `os.path.expanduser` twice for `auth_token` and twice for `client_secret` right? So just move those two lines to post_config_hook, and then I can use them with the path already expanded in `_get_credentials()`, right?

Reply:
Right. :-)  Put them with other `self`.